Add PEP 712 converters to the `dataclasses` spec (#1783)
python/typing: docs/spec/dataclasses.rst

* ``alias`` is an optional str parameter that provides an alternative
  name for the field. This alternative name is used in the synthesized
  ``__init__`` method.

> * ``converter`` is an optional parameter that specifies a single-argument

Reviewer: Should say "single-parameter", not "single-argument". Callables have parameters; argument are passed to the callable.

Author: ```suggestion
* ``converter`` is an optional parameter that specifies a single-parameter
```

>   callable used to convert values when assigning to the associated

Reviewer: ```suggestion
* ``converter`` is an optional parameter that specifies a single-parameter
```

>   attribute (Including the assignment in ``__init__``). The parameter

Reviewer: Nit: "Including" should not be capitalized here.

>   type of this single argument provides the type of the synthesized

Reviewer: "The parameter type of this single argument" is confusing. Recommend simplifying to "The the of this parameter".

>   attribute (Including the assignment in ``__init__``). The parameter
>   type of this single argument provides the type of the synthesized

Author: ```suggestion
  attribute (including the assignment in ``__init__``). The type
  of this argument provides the type of the synthesized
```

* ``alias`` is an optional str parameter that provides an alternative
  name for the field. This alternative name is used in the synthesized
  ``__init__`` method.
* ``converter`` is an optional parameter that specifies a

>   callable used to convert values when assigning to the associated
>   attribute.

Reviewer 2: Wording nit: "associated attribute" feels like very specific terminology, but it's not used in describing any of the other field specifier parameters. If there is something particularly unique to `converter` that we are trying to communicate with this wording, maybe we should communicate it more explicitly? Otherwise, it looks more consistent to just use "field" rather than "associated attribute."
```suggestion
  callable used to convert values when assigning to the field.
```

  dataclass must be explicitly annotated as e.g. ``x: ClassVar[Final[int]] =
  3``.

Converters
^^^^^^^^^^

The ``converter`` parameter can be specified in a field definition to provide a callable

> used to convert values when assigning to the associated attribute. This feature allows for automatic type

Reviewer 2: Here again, is "associated attribute" meaningfully different from just saying "field"?

conversion and validation during attribute assignment.

Converter behavior:

* The converter is used for all attribute assignment, including assignment
  of default values, assignment in synthesized ``__init__`` methods
  and direct attribute setting (e.g., ``obj.attr = value``).
* The converter is not used when reading attributes, as the attributes should already have been converted.

Typing rules for converters:

* The ``converter`` must be a callable that must accept a single positional argument

>   (but may optionally accept other arguments, which are ignored for typing purposes).

Reviewer 3: Provided the other arguments are optional (i.e., they have a default or are variadic). Hopefully that's clear but maybe we can use a wording that makes it more explicit.

* The type of the first positional parameter provides the type of the synthesized ``__init__`` parameter
  associated with the field.
* The return type of the callable must be assignable to the field's declared type.
* If ``default`` or ``default_factory`` are provided, the type of the default value should be

>   assignable to the first positional parameter of the ``converter``

Reviewer 4: grammar nit: missing period at the end of the sentence

Author: ```suggestion
  assignable to the first positional parameter of the ``converter``.
```

Example usage:

.. code-block:: python

    def str_or_none(x: Any) -> str | None:
        return str(x) if x is not None else None

    @custom_dataclass
    class Example:
        int_field: int = field(converter=int)

>         str_field: str | None = field(converter=str_or_none)

Reviewer 5: since `dataclasses.field` isn't going to gain the `converter=` kwarg: in your example can you show the definition of `field` as a `field_specifier` of `custom_dataclass` to show that this is not `dataclasses.field`?

Author: How about `custom_field`?

  dataclass must be explicitly annotated as e.g. ``x: ClassVar[Final[int]] =
  3``.

Converters
^^^^^^^^^^

> The ``converter`` parameter can be specified in a field definition to provide a single-argument callable

Reviewer: "single-argument" -> "single-parameter"

Author: ```suggestion
The ``converter`` parameter can be specified in a field definition to provide a single-parameter callable
```

used to convert values when assigning to the associated attribute. This feature allows for automatic type
conversion and validation during object initialization and attribute assignment.

Converter behavior:

> * For non-frozen dataclasses, the converter is used for all attribute assignment, including assignment

Reviewer 4: I find it a little confusing for this section to refer to "non-frozen dataclasses," since that sounds to me like you're talking about stdlib dataclasses. But stdlib dataclasses don't have the `converter` parameter, right?

Maybe you could write this as "non-frozen classes impacted by `dataclass_transform`" and then just "frozen classes" or "non-frozen classes" in the rest of this section.

> * For non-frozen dataclasses, the converter is used for all attribute assignment, including assignment

Reviewer: I agree with rchen152 that this section is confusing. Why does the behavior for frozen classes need to be discussed at all here? A converter should work anywhere that assignment works — including initial assignment via the synthesized `__init__`. If a class is frozen, then subsequent assignments will generate an exception. That's how frozen classes work, so I don't think there's a need to explain that here.

Author: The intent was to ensure there's no ambiguity on frozen dataclasses, since one could mistakenly assume they don't "assign" their attributes.
I think here we'll just say "both frozen and non-frozen" and move on...

  of default values and direct attribute setting (e.g., ``obj.attr = value``).
* For frozen dataclasses, the converter is only used inside a ``dataclass``-synthesized ``__init__``
  when setting the attribute.
* The converter is not used when reading attributes, as the attributes should already have been converted.

Typing rules for converters:

> * The ``converter`` must be a callable that accepts a single positional argument.

Reviewer: "callable that accepts a single positional argument" => "a callable with one positional-only parameter"

Author: Actually I'll push back on this one, since the callable can be more than a one-positional-parameter-only callable. E.g. `foo(*args)` or `foo(x, y=None)` should be compatible.

Author: (although I think this is more of a problem _most_ places we say single-parameter. Maybe it's worth clarifying once at the top, that by "single-parameter" we mean "single-parameter-compatible"? (or just use that slightly verbose wording throughout?)

Reviewer 2: Let's stay away from "compatible" -- that word isn't defined to have any particular meaning in the spec.

I think the current wording here ("callable that accepts a single positional argument") is OK, and accurately describes the requirements, without placing additional unnecessary restrictions on the callable.

Unless @erictraut you were intending to suggest a more substantive change, and you believe the callable really should be restricted to only accepting one parameter, and that parameter must be positional-only? I'm not sure I see the rationale for those limitations.

Reviewer: Above, the spec talks about a "single-parameter callable", so I was trying to make the language consistent here and use "parameter" rather than "argument", but in the process I unintentionally added a new limitation. I agree that limiting it to a positional-only parameter is unnecessary. The current wording is fine.

Author: Yeah this is exactly the kind of ambiguity I'm hoping to avoid by being explicit 😂

I'm happy to tweak the wording though. Would "assignable" be a correct thing to say?

Alternatively something akin to "must accept a single positional argument" (which allows any parameter spec that matches the criteria)

> * The parameter type of this single argument provides the type of the synthesized ``__init__`` parameter

Reviewer: "The parameter type of this single argument" => "The type of the parameter"

Author: ```suggestion
* The parameter type of the parameter provides the type of the synthesized ``__init__`` parameter
```

  associated with the field.
* The return type of the callable must be compatible with the field's declared type.

> When used with ``default`` or ``default_factory``:

Reviewer 2: This seems redundant with the text of the bullet point itself; could just eliminate this and include the next bullet point in the above list of typing rules directly?

Author: Good eye

* If a ``converter`` is provided alongside ``default`` or ``default_factory``, the type of the default

>   value should be assignable to the type of the single argument to the ``converter`` callable.

Reviewer: "the type of the single argument to the ``converter`` callable" => "the ``converter`` parameter"

Author: ```suggestion
  value should be assignable to the the ``converter`` parameter.
```

> * Default values are unconditionally converted using the ``converter`` if provided.

Reviewer: I recommend moving this point up to the first sentence, as in:

"If a ``converter`` is provided alongside ``default`` or ``default_factory``, default values are converted using the ``converter``".

Then you can explain how a type checker should validate this behavior.

> * Default values are unconditionally converted using the ``converter`` if provided.

Reviewer: What does "unconditionally" mean here? Perhaps just delete that word if it's not providing any meaning.

Example usage:

.. code-block:: python

    def str_or_none(x: Any) -> str | None:
        return str(x) if x is not None else None

>     @dataclasses.dataclass

Reviewer 4: Similar comment to above - rather than using stdlib dataclasses, I think it would be better to cook up a little example using `dataclass_transform`.

Author: Oh this was just a copy paste mistake 😅

    class Example:

>         int_field: int = dataclasses.field(converter=int)

Reviewer: The `dataclasses.field` doesn't support converters. I recommend changing this to simply `field`, which the reader can assume is some field specifier used with a dataclass transform.
